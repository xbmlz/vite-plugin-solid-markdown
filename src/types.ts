import type { PluggableList } from '@mdx-js/mdx/lib/core'
import type { FilterPattern } from 'vite'

export interface Options {
  /**
   * Class names for wrapper div
   *
   * @default 'markdown-body'
   */
  wrapperClasses?: string | string[]

  /**
   * Component name to wrapper with
   *
   * @default undefined
   */
  wrapperComponent?: string | undefined | null

  /**
   * Remark plugins to use
   *
   * @default []
   */
  remarkPlugins?: PluggableList

  /**
   * Rehype plugins to use
   *
   * @default []
   */
  rehypePlugins?: PluggableList

  include?: FilterPattern
  exclude?: FilterPattern
}

export interface ResolvedOptions extends Required<Options> {
  wrapperClasses: string
}
