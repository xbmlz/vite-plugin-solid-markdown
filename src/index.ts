import { compileSync } from '@mdx-js/mdx'
import type { Plugin } from 'vite'
import { createFilter } from 'vite'

function VitePluginSolidMarkdown(): Plugin {
  // mdx md files
  const filter = createFilter('**/*.mdx', '**/node_modules/**')
  return {
    name: 'vite-plugin-solid-markdown',
    enforce: 'pre',
    transform(raw: string, id: string) {
      if (!filter(id)) return
      const code = compileSync(raw, {
        jsx: true,
        jsxImportSource: 'solid-js',
        providerImportSource: 'solid-mdx',
      })
      return {
        code: String(code),
        map: { mappings: '' } as any,
      }
    },
  }
}
export default VitePluginSolidMarkdown
