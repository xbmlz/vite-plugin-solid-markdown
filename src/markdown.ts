import { compileSync } from '@mdx-js/mdx'
import type { TransformResult } from 'vite'
import type { ResolvedOptions } from './types'

export function createMarkdown(options: ResolvedOptions) {
  function rehypeWrapperClasses() {
    // unknown is not assignable to type Element
    return function (tree: any) {
      if (options.wrapperClasses) {
        const wrapper = {
          type: 'element',
          tagName: 'div',
          properties: {
            className: options.wrapperClasses,
          },
        }
        // @ts-expect-error - property children does not exist on type Element
        wrapper.children = tree.children
        tree.children = [wrapper]
      }
      if (options.wrapperComponent) {
        const wrapper = {
          type: 'element',
          tagName: options.wrapperComponent,
          properties: {},
        }
        // @ts-expect-error - property children does not exist on type Element
        wrapper.children = tree.children
        tree.children = [wrapper]
      }
    }
  }

  return (id: string, raw: string): TransformResult => {
    const code = String(
      compileSync(raw, {
        jsx: true,
        jsxImportSource: 'solid-js',
        providerImportSource: 'solid-mdx',
        rehypePlugins: [...options.rehypePlugins, rehypeWrapperClasses],
        remarkPlugins: [...options.remarkPlugins],
      })
    )

    return {
      code,
      map: { mappings: '' } as any,
    }
  }
}
