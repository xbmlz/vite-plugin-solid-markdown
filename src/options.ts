import { toArray } from '@antfu/utils'
import type { Options, ResolvedOptions } from './types'

export function resolveOptions(userOptions: Options): ResolvedOptions {
  const defaultOptions: ResolvedOptions = {
    wrapperClasses: 'markdown-body',
    wrapperComponent: null,
    remarkPlugins: [],
    rehypePlugins: [],
    include: null,
    exclude: null,
  }

  const options = { ...defaultOptions, ...userOptions }

  options.wrapperClasses = toArray(options.wrapperClasses)
    .filter((i?: string) => i)
    .join(' ')

  return options as ResolvedOptions
}
