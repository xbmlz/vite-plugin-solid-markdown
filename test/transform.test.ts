import { describe, expect, it } from 'vitest'
import { createMarkdown } from '../src/markdown'
import { resolveOptions } from '../src/options'

describe('transform', () => {
  const options = resolveOptions({})
  const markdownToVue = createMarkdown(options)

  it('basic', async () => {
    const md = `

    # Hello

    - A
    - B
    - C
    `
    expect((await markdownToVue('', md)).code).toMatchSnapshot()
  })

  it('jsx', async () => {
    const md = `

    import Counter from '~/components/Counter'

    <Counter />
    `
    expect((await markdownToVue('', md)).code).toMatchSnapshot()
  })

  it('frontmatter', async () => {
    const md = `

    export const name = 'My Cool App'
    export const title = 'Hello ' + name + '!'

    This is {name}
    `
    expect((await markdownToVue('', md)).code).toMatchSnapshot()
  })
})
