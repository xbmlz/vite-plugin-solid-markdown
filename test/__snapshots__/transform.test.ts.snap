// Vitest Snapshot v1

exports[`transform > basic 1`] = `
"/*@jsxRuntime automatic @jsxImportSource solid-js*/
import {useMDXComponents as _provideComponents} from \\"solid-mdx\\";
function _createMdxContent(props) {
  const _components = Object.assign({
    div: \\"div\\",
    h1: \\"h1\\",
    ul: \\"ul\\",
    li: \\"li\\"
  }, _provideComponents(), props.components);
  return <_components.div className=\\"markdown-body\\"><_components.h1>{\\"Hello\\"}</_components.h1>{\\"\\\\n\\"}<_components.ul>{\\"\\\\n\\"}<_components.li>{\\"A\\"}</_components.li>{\\"\\\\n\\"}<_components.li>{\\"B\\"}</_components.li>{\\"\\\\n\\"}<_components.li>{\\"C\\"}</_components.li>{\\"\\\\n\\"}</_components.ul></_components.div>;
}
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = Object.assign({}, _provideComponents(), props.components);
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
}
export default MDXContent;
"
`;

exports[`transform > frontmatter 1`] = `
"/*@jsxRuntime automatic @jsxImportSource solid-js*/
import {useMDXComponents as _provideComponents} from \\"solid-mdx\\";
function _createMdxContent(props) {
  const _components = Object.assign({
    div: \\"div\\",
    p: \\"p\\"
  }, _provideComponents(), props.components);
  return <_components.div className=\\"markdown-body\\"><_components.p>{\\"export const name = 'My Cool App'\\\\nexport const title = 'Hello ' + name + '!'\\"}</_components.p>{\\"\\\\n\\"}<_components.p>{\\"This is \\"}{name}</_components.p></_components.div>;
}
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = Object.assign({}, _provideComponents(), props.components);
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
}
export default MDXContent;
"
`;

exports[`transform > jsx 1`] = `
"/*@jsxRuntime automatic @jsxImportSource solid-js*/
import {useMDXComponents as _provideComponents} from \\"solid-mdx\\";
function _createMdxContent(props) {
  const _components = Object.assign({
    div: \\"div\\",
    p: \\"p\\"
  }, _provideComponents(), props.components), {Counter} = _components;
  if (!Counter) _missingMdxReference(\\"Counter\\", true);
  return <_components.div className=\\"markdown-body\\"><_components.p>{\\"import Counter from '~/components/Counter'\\"}</_components.p>{\\"\\\\n\\"}<Counter /></_components.div>;
}
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = Object.assign({}, _provideComponents(), props.components);
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent {...props} /></MDXLayout> : _createMdxContent(props);
}
export default MDXContent;
function _missingMdxReference(id, component) {
  throw new Error(\\"Expected \\" + (component ? \\"component\\" : \\"object\\") + \\" \`\\" + id + \\"\` to be defined: you likely forgot to import, pass, or provide it.\\");
}
"
`;
